bc_obps/compliance/service/compliance_invoice_service.py

```diff
@@ -17,58 +26,231 @@ def generate_invoice_pdf(
+            errors["missing_reporting_year"] = msg
+            return {"errors": errors}
+
+        # Find the ComplianceChargeRate for that reporting_year
+        try:
+            charge_rate_obj = ComplianceChargeRate.objects.get(reporting_year=reporting_year_obj)
+            charge_rate_decimal: Decimal = charge_rate_obj.rate
+        except ObjectDoesNotExist:
+            msg = f"No ComplianceChargeRate found for ReportingYear " f"{context_reporting_year}."
+            errors["missing_charge_rate"] = msg
+            return {"errors": errors}
+        except Exception as exc:
+            msg = f"Unexpected error fetching ComplianceChargeRate: {exc}"
+            errors["charge_rate_error"] = msg
+            return {"errors": errors}
+
+        # --------------------------------------------------------------------
+        # 3) Compute equivalent_amount = (fee_amount_dollars × charge_rate)
+        # --------------------------------------------------------------------
+        try:
+            base_fee: Decimal = compliance_obligation.fee_amount_dollars or Decimal("0.00")
+            raw_equivalent = base_fee * charge_rate_decimal
+            equivalent_amount_decimal = raw_equivalent.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
+            compliance_obligation_equivalent_amount = f"${equivalent_amount_decimal:,.2f}"
+            compliance_obligation_charge_rate = f"${charge_rate_decimal:,.2f} per tCO₂e"
+        except Exception as exc:
+            msg = f"Failed to compute equivalent amount: {exc}"
+            errors["compute_equivalent_error"] = msg
+            return {"errors": errors}
+
+        # --------------------------------------------------------------------
+        # 4) Build remaining invoice fields
+        # --------------------------------------------------------------------
+        try:
+            invoice_number = "OBI+six-digits"  # replace with real ID if you have one
+
+            if compliance_obligation.fee_date:
+                invoice_date_obj = compliance_obligation.fee_date
+            else:
+                if compliance_obligation.created_at:
+                    invoice_date_obj = compliance_obligation.created_at.date()
+                else:
+                    raise ValueError(
+                        f"Obligation (ID {getattr(compliance_obligation, 'pk', 'unknown')}) "
+                        "has neither fee_date nor created_at."
+                    )
+            invoice_date = invoice_date_obj.strftime("%b %-d, %Y")
+            invoice_due_date = compliance_obligation.obligation_deadline.strftime("%b %-d, %Y")
+            invoice_printed_date = timezone.now().strftime("%b %-d, %Y")
+
+            operator_name = operator.legal_name
+            operator_address_line1 = operator_address.street_address
+            operator_address_line2 = (
+                f"{operator_address.municipality}, " f"{operator_address.province}, " f"{operator_address.postal_code}"
```

Comment:
one fstring should be enough to form this address.
`f"{operator_address.municipality}, {operator_address.province}, {operator_address.postal_code}"`